File: 3307. Find the K-th Character in String Game II/find_the_kth_character_in_string_game_II.py
#Brute-Force
#T.C.->O(2^n)
#S.C.->O(2^n)
class Solution:
    def kthCharacter(self, k: int, operations: List[int]) -> str:
        s = "a"
        for op in operations:
            if op == 0:
                s += s
            else:
                s += ''.join(chr((ord(c) - 96) % 26 + 97) if c != 'z' else 'a' for c in s)
        return s[k - 1]

#Optimal
#T.C.->O(n)
#S.C.->O(1)
class Solution:
    def kthCharacter(self, k: int, operations: List[int]) -> str:
        shift = 0
        k -= 1
        for i in range(len(operations)-1, -1, -1):
            half = float('inf') if i >= 60 else 1 << i
            if k >= half:
                k -= half
                shift += operations[i]
        return chr(ord('a') + shift % 26)

File: 3307. Find the K-th Character in String Game II/find_the_kth_character_in_string_game_II.cpp
//Brute-Force
//T.C.->O(2^n)
//S.C.->O(2^n)
class Solution {
public:
    char kthCharacter(long long k, vector<int>& operations) {
        string s = "a";
        for (int op : operations) {
            string t;
            if (op == 0) 
                t = s;
            else
                for (char c : s) 
                    t += (c == 'z' ? 'a' : c + 1);
            s += t;
        }
        return s[k - 1];
    }
};

//Optimal
//T.C.->O(n)
//S.C.->O(1)
class Solution {
public:
    char kthCharacter(long long k, vector<int>& operations) {
        int shift = 0;
        k--;
        
        for(int i = operations.size() - 1; i >= 0; i--) {
            long long half;
            if(i >= 60) {
                half = LLONG_MAX;
            } else {
                half = 1LL << i;
            }
            
            if(k >= half) {
                k -= half;
                shift += operations[i];
            }
        }
        
        return (char)('a' + shift % 26);
    }
};
